File: diagnostics/voge-nav-bridge/app/src/main/java/it/motolink/vogenavbridge/VogeNavPacketEncoder.java
package it.motolink.vogenavbridge;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public final class VogeNavPacketEncoder {
    public static final int CMD_NAV = 0x6A;
    public static final int CMD_NEXT_1 = 0x6B;
    public static final int CMD_NEXT_2 = 0x6C;
    public static final int CMD_CUR_1 = 0x6D;
    public static final int CMD_CUR_2 = 0x6E;

    private VogeNavPacketEncoder() {}

    public static NavModel romeTestModel() {
        NavModel m = new NavModel();
        m.routeRemainDistM = 2400;
        m.routeRemainTimeSec = 600;
        m.arriveRemainSec = 600;
        m.curRoadRemainDistM = 300;
        m.curRoadName = "Via Nazionale";
        m.nextRoadDirection = 5; // VOGE: right turn
        m.nextRoadName = "Via del Corso";
        m.roadFlag = 0;
        m.naviOnOff = 0;
        m.annularDegrees = 0;
        m.realDistance = true;
        m.sourceTitle = "TEST ROMA: svolta a destra verso Via del Corso";
        m.sourceSubText = "Dati sintetici, non provenienti da Maps";
        return m;
    }

    public static List<byte[]> allFrames(NavModel m) {
        List<byte[]> out = new ArrayList<>();
        out.add(navFrame(m));
        out.add(textFrame(CMD_NEXT_1, m.nextRoadName, false));
        out.add(textFrame(CMD_NEXT_2, m.nextRoadName, true));
        out.add(textFrame(CMD_CUR_1, m.curRoadName, false));
        out.add(textFrame(CMD_CUR_2, m.curRoadName, true));
        return out;
    }

    public static byte[] navFrame(NavModel m) {
        byte[] p = base(CMD_NAV);
        putBe(p, 2, clamp(m.curRoadRemainDistM, 0, 0xFFFFFF), 3);
        putBe(p, 5, clamp(m.routeRemainTimeSec, 0, 0xFFFF), 2);
        putBe(p, 7, clamp(m.routeRemainDistM, 0, 0xFFFFFF), 3);
        p[10] = (byte) clamp(m.nextRoadDirection, 0, 255);

        long deltaSec = m.arriveRemainSec > 0 ? m.arriveRemainSec : m.routeRemainTimeSec;
        long arriveMs = System.currentTimeMillis() + Math.max(0, deltaSec) * 1000L;
        Calendar cal = Calendar.getInstance();
        cal.setTimeInMillis(arriveMs);
        int hour24 = cal.get(Calendar.HOUR_OF_DAY);
        p[11] = (byte) hour24;
        p[12] = (byte) cal.get(Calendar.MINUTE);
        p[13] = (byte) (hour24 >= 12 ? 1 : 0);
        p[14] = (byte) clamp(m.roadFlag, 0, 255);
        p[15] = (byte) clamp(m.naviOnOff, 0, 255);
        p[16] = (byte) clamp(m.annularDegrees, 0, 255);
        return checksum(p);
    }

    public static byte[] textFrame(int command, String text, boolean secondChunk) {
        byte[] p = base(command);
        p[2] = (byte) ("zh".equalsIgnoreCase(Locale.getDefault().getLanguage()) ? 1 : 0);
        byte[] utf = (text == null ? "" : text).getBytes(StandardCharsets.UTF_8);
        int src = secondChunk ? 15 : 0;
        int count = Math.min(15, Math.max(0, utf.length - src));
        if (count > 0) System.arraycopy(utf, src, p, 3, count);
        return checksum(p);
    }

    public static String directionName(int code) {
        switch (code) {
            case 1: return "STRAIGHT";
            case 2: return "UTURN_LEFT";
            case 3: return "UTURN_RIGHT";
            case 4: return "LEFT";
            case 5: return "RIGHT";
            case 6: return "SLIGHT_LEFT";
            case 7: return "SLIGHT_RIGHT";
            case 8: return "SHARP_LEFT";
            case 9: return "SHARP_RIGHT";
            default: return "NONE/UNKNOWN";
        }
    }

    private static byte[] base(int command) {
        byte[] p = new byte[20];
        p[0] = 0x01;
        p[1] = (byte) command;
        p[19] = 0x04;
        return p;
    }

    private static void putBe(byte[] dest, int offset, int value, int count) {
        for (int i = 0; i < count; i++) {
            int shift = (count - 1 - i) * 8;
            dest[offset + i] = (byte) ((value >> shift) & 0xFF);
        }
    }

    private static byte[] checksum(byte[] p) {
        p[p.length - 2] = 0;
        int x = 0;
        for (byte b : p) x ^= (b & 0xFF);
        p[p.length - 2] = (byte) x;
        return p;
    }

    private static int clamp(int v, int min, int max) {
        return Math.max(min, Math.min(max, v));
    }
}

File: app/src/main/java/it/motolink/app/VogeNavPacketEncoder.kt
package it.motolink.app

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import java.util.Calendar
import java.util.Locale

object VogeNavPacketEncoder {
    const val CMD_HEARTBEAT = 0x5A
    const val CMD_NAV = 0x6A
    const val CMD_NEXT_1 = 0x6B
    const val CMD_NEXT_2 = 0x6C
    const val CMD_CUR_1 = 0x6D
    const val CMD_CUR_2 = 0x6E

    fun heartbeatFrame(context: Context): ByteArray {
        val p = base(CMD_HEARTBEAT)
        val battery = runCatching {
            val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            intent?.getIntExtra("level", 0) ?: 0
        }.getOrDefault(0).coerceIn(0, 255)
        val now = Calendar.getInstance()
        val hour24 = now.get(Calendar.HOUR_OF_DAY)
        p[2] = 0
        p[3] = battery.toByte()
        p[4] = 0x32
        p[5] = hour24.toByte()
        p[6] = now.get(Calendar.MINUTE).toByte()
        p[7] = now.get(Calendar.SECOND).toByte()
        p[8] = if (hour24 >= 12) 1 else 0
        p[9] = 0
        p[10] = 0x02
        return checksum(p)
    }

    fun allFrames(m: VogeNavModel): List<ByteArray> = listOf(
        navFrame(m),
        textFrame(CMD_NEXT_1, m.nextRoadName, false),
        textFrame(CMD_NEXT_2, m.nextRoadName, true),
        textFrame(CMD_CUR_1, m.curRoadName, false),
        textFrame(CMD_CUR_2, m.curRoadName, true)
    )

    fun navFrame(m: VogeNavModel): ByteArray {
        val p = base(CMD_NAV)
        putBe(p, 2, m.curRoadRemainDistM.coerceIn(0, 0xFFFFFF), 3)
        putBe(p, 5, m.routeRemainTimeSec.coerceIn(0, 0xFFFF), 2)
        putBe(p, 7, m.routeRemainDistM.coerceIn(0, 0xFFFFFF), 3)
        p[10] = m.nextRoadDirection.coerceIn(0, 255).toByte()
        val deltaSec = if (m.arriveRemainSec > 0) m.arriveRemainSec else m.routeRemainTimeSec
        val cal = Calendar.getInstance().apply {
            timeInMillis = System.currentTimeMillis() + deltaSec.coerceAtLeast(0).toLong() * 1000L
        }
        val hour24 = cal.get(Calendar.HOUR_OF_DAY)
        p[11] = hour24.toByte()
        p[12] = cal.get(Calendar.MINUTE).toByte()
        p[13] = if (hour24 >= 12) 1 else 0
        p[14] = m.roadFlag.coerceIn(0, 255).toByte()
        p[15] = m.naviOnOff.coerceIn(0, 255).toByte()
        p[16] = m.annularDegrees.coerceIn(0, 255).toByte()
        return checksum(p)
    }

    private fun textFrame(command: Int, text: String?, secondChunk: Boolean): ByteArray {
        val p = base(command)
        p[2] = if (Locale.getDefault().language.equals("zh", true)) 1 else 0
        val utf = (text ?: "").toByteArray(Charsets.UTF_8)
        val src = if (secondChunk) 15 else 0
        val count = minOf(15, (utf.size - src).coerceAtLeast(0))
        if (count > 0) utf.copyInto(p, destinationOffset = 3, startIndex = src, endIndex = src + count)
        return checksum(p)
    }

    private fun base(command: Int): ByteArray = ByteArray(20).also {
        it[0] = 0x01
        it[1] = command.toByte()
        it[19] = 0x04
    }

    private fun putBe(dest: ByteArray, offset: Int, value: Int, count: Int) {
        for (i in 0 until count) {
            val shift = (count - 1 - i) * 8
            dest[offset + i] = ((value shr shift) and 0xFF).toByte()
        }
    }

    private fun checksum(p: ByteArray): ByteArray {
        p[p.lastIndex - 1] = 0
        var x = 0
        p.forEach { x = x xor (it.toInt() and 0xFF) }
        p[p.lastIndex - 1] = x.toByte()
        return p
    }
}
